### PR TITLE
upgrade rustpython-parser (not going to be finished)

### PR DESCRIPTION
**Pull request by youknowone**
I didn't finish it but I made a few changes as example.

You may want to rewrite ASTConverter using `rustpython_ast::Fold` trait.

Please feel free to request features you need from parser/ast side.

### CODE REVIEW

File: crates/py2erg/convert.rs [hunk 1]
```diff
@@ -327,15 +327,15 @@ impl ASTConverter {
         let token = Token::new(
             TokenKind::Symbol,
             cont,
-            loc.row() as u32,
-            loc.column() as u32 - 1,
+            loc.row.get(),
+            loc.column.get() - 1,
```
Comment (youknowone) [inline crates/py2erg/convert.rs:331]:
```suggestion
            loc.column.to_zero_indexed(),
```

**Comment by youknowone** [comment]
@mtshiba Do you need any help?

**Comment by mtshiba** [comment]
No, it's okay. I'll complete it over the weekend as there are so many changes.

**Comment by mtshiba** [comment]
@youknowone Please help me, how can I call `parse`?

https://github.com/mtshiba/pylyzer/blob/39c27524ec490ab86ec887604dcfd8ce802003ab/src/analyze.rs#L18-L26

```console
no function or associated item named `parse` found for struct `rustpython_ast::ModModule` in the current scope
```

`rustpython_ast::located::ModModule` does not seem to implement `Parse`.
I changed `rustpython_ast::located::ModModule` to `rustpython_ast::ModModule` then I got

```console
mismatched types
expected struct `rustpython_ast::ModModule<rustpython_parser::source_code::SourceRange>`
   found struct `rustpython_ast::ModModule<rustpython_parser::ruff_text_size::TextRange>`
```

**Comment by youknowone** [comment]
Hi! you have to use `Parse for rustpython_ast::ModModule` first. And transform the raw AST to located one later using `SourceLocator`, which is recently replaced by `LinearLocator`.
Probably I'd better to add `Parse for rustpython_ast::located::ModModule` for convenience.

**Comment by mtshiba** [comment]
@youknowone The parser crashed when folding a function def with default parameters. Is this a bug?

```rust
use rustpython_ast::{Fold, ModModule, source_code::LinearLocator};
use rustpython_parser::Parse;

fn main() {
    let code = "def f(x=1): pass";
    let module = ModModule::parse(code, "_").unwrap();
    println!("{module:?}");
    let mut locator = LinearLocator::new(code);
    let module = locator.fold(module).unwrap();
    println!("{module:?}");
}
```

```console
ModModule { range: 0..16, body: [FunctionDef(StmtFunctionDef { range: 0..16, name: Identifier("f"), args: Arguments { range: 6..9, posonlyargs: [], args: [ArgWithDefault { range: 6..7, def: Arg { range: 6..7, arg: Identifier("x"), annotation: None, type_comment: None }, default: Some(Constant(ExprConstant { range: 8..9, value: Int(1), kind: None })) }], vararg: None, kwonlyargs: [], kwarg: None }, body: [Pass(StmtPass { range: 12..16 })], decorator_list: [], returns: None, type_comment: None })], type_ignores: [] }
thread 'main' panicked at '9 -> 7 =1', .../checkouts/parser-aad50aa3362d3a85/b2f95e2/core/src/source_code.rs:139:9
```

**Comment by youknowone** [comment]
Oh, I will look in it. Thank you! You can use `RandomLocator` instead of `LinearLocator`, which is slower but more stable version.

**Comment by mtshiba** [comment]
Thanks @youknowone!

**Comment by youknowone** [comment]
I am happy to see that 😄

**Comment by youknowone** [comment]
The LinearLocator panic is not reproducible. It may used older unfixed version.

**Pull request merged**